### PR TITLE
Improve robust for backup huge data size

### PR DESCRIPTION
**Pull request by zhoney**
1. add timeout set
2. add splitsize set
3. add 80w exceed process
4. add HDFS support

Change-Id: I2976270f8d8e6970f5966f48a5d91e9ab7056ce1

### CODE REVIEW

File: src/main/java/com/baidu/hugegraph/manager/RestoreManager.java [hunk 1]
```diff
@@ -240,35 +239,43 @@ private void restore(HugeType type, File file,
...
-                throw new SerializeException(
-                          "Can't find value of the key: %s in json.", key);
-            } else {
-                JavaType t = mapper.getTypeFactory()
-                                   .constructParametricType(List.class, clazz);
-                return (List<T>) mapper.readValue(element.toString(), t);
+    private void restoreZip(HugeType type, File file,
+                            BiConsumer<String, String> consumer) {
+        E.checkArgument(
+                file.exists() && file.isFile() && file.canRead(),
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/RestoreManager.java]: align

File: src/main/java/com/baidu/hugegraph/manager/RestoreManager.java [hunk 2]
```diff
@@ -305,4 +312,9 @@ private Object updateVid(String label, Object id) {
         }
         return id;
     }
+
+    private void printRestoreFiles(HugeType type, List<File> files) {
+        Printer.print("Restoring %s ...", type);
+        Printer.printList("files", files);
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/RestoreManager.java]: prefer to add two spaces before "files"

**Review by javeme — CHANGES_REQUESTED** [review]

File: src/main/java/com/baidu/hugegraph/manager/DumpGraphManager.java [hunk 3]
```diff
@@ -43,12 +43,13 @@
...
-    public DumpGraphManager(String url, String graph) {
-        this(url, graph, "JsonFormatter");
+    public DumpGraphManager(String url, String graph, Integer timeout) {
+        this(url, graph, "JsonFormatter", timeout);
     }
 
-    public DumpGraphManager(String url, String graph, String formatter) {
-        super(url, graph);
+    public DumpGraphManager(String url, String graph,
+                            String formatter, Integer timeout) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/DumpGraphManager.java]: put parameter timeout before parameter formatter

File: src/main/java/com/baidu/hugegraph/base/ToolManager.java [hunk 4]
```diff
@@ -19,23 +19,53 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper;
+
 public class ToolManager {
 
     protected final ToolClient client;
     private final String type;
 
-    public ToolManager(String url, String graph, String type) {
-        this.client = new ToolClient(url, graph);
+    public ToolManager(String url, String graph, String type, Integer timeout) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/base/ToolManager.java]: int timeout

File: src/main/java/com/baidu/hugegraph/manager/RestoreManager.java [hunk 5]
```diff
@@ -240,35 +249,43 @@ private void restore(HugeType type, File file,
...
-                throw new SerializeException(
-                          "Can't find value of the key: %s in json.", key);
-            } else {
-                JavaType t = mapper.getTypeFactory()
-                                   .constructParametricType(List.class, clazz);
-                return (List<T>) mapper.readValue(element.toString(), t);
+    private void restoreZip(HugeType type, File file,
+                            BiConsumer<String, String> consumer) {
+        E.checkArgument(
+                file.exists() && file.isFile() && file.canRead(),
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/RestoreManager.java]: address this comments

File: src/main/java/com/baidu/hugegraph/manager/RestoreManager.java [hunk 6]
```diff
@@ -305,4 +322,9 @@ private Object updateVid(String label, Object id) {
         }
         return id;
     }
+
+    private void printRestoreFiles(HugeType type, List<File> files) {
+        Printer.print("Restoring %s ...", type);
+        Printer.printList("files", files);
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/RestoreManager.java]: ditto

**Review by javeme — CHANGES_REQUESTED** [review]

File: src/main/java/com/baidu/hugegraph/base/ToolClient.java [hunk 7]
```diff
@@ -32,14 +32,14 @@
     private HugeClient client;
     private ObjectMapper mapper;
 
-    public ToolClient(String url, String graph) {
-        this.client = new HugeClient(url, graph);
+    public ToolClient(String url, String graph, Integer timeout) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/base/ToolClient.java]: int timeout

File: src/main/java/com/baidu/hugegraph/base/ToolClient.java [hunk 8]
```diff
@@ -32,14 +32,14 @@
...
+        this.client = new HugeClient(url, graph, timeout);
         this.mapper = new ObjectMapper();
     }
 
     public ToolClient(String url, String graph,
-                      String username, String password) {
-        this.client = new HugeClient(url, graph, username, password);
+                      String username, String password, Integer timeout) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/base/ToolClient.java]: ditto

File: src/main/java/com/baidu/hugegraph/cmd/HugeGraphCommand.java [hunk 9]
```diff
@@ -110,9 +121,12 @@ private void execute(String subCmd, JCommander jCommander) {
             case "backup":
                 Printer.print("Graph '%s' start backup!", this.graph());
                 SubCommands.Backup backup = this.subCommand(subCmd);
+                this.timeout(120);
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/cmd/HugeGraphCommand.java]: prefer to add a parameter

File: src/main/java/com/baidu/hugegraph/manager/DumpGraphManager.java [hunk 10]
```diff
@@ -43,12 +43,13 @@
 
     private Formatter dumpFormatter;
 
-    public DumpGraphManager(String url, String graph) {
-        this(url, graph, "JsonFormatter");
+    public DumpGraphManager(String url, String graph, Integer timeout) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/DumpGraphManager.java]: ditto

File: src/main/java/com/baidu/hugegraph/manager/DumpGraphManager.java [hunk 11]
```diff
@@ -43,12 +43,13 @@
...
+        this(url, graph, timeout, "JsonFormatter");
     }
 
-    public DumpGraphManager(String url, String graph, String formatter) {
-        super(url, graph);
+    public DumpGraphManager(String url, String graph,
+                            Integer timeout, String formatter) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/DumpGraphManager.java]: ditto

File: src/main/java/com/baidu/hugegraph/manager/GremlinManager.java [hunk 12]
```diff
@@ -27,13 +27,13 @@
 
 public class GremlinManager extends ToolManager {
 
-    public GremlinManager(String url, String graph) {
-        super(url, graph, "gremlin");
+    public GremlinManager(String url, String graph, Integer timeout) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/GremlinManager.java]: ditto

File: src/main/java/com/baidu/hugegraph/manager/GremlinManager.java [hunk 13]
```diff
@@ -27,13 +27,13 @@
...
+        super(url, graph, "gremlin", timeout);
     }
 
     public GremlinManager(String url, String graph,
-                          String username, String password) {
-        super(url, graph, username, password, "gremlin");
+                          String username, String password, Integer timeout) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/GremlinManager.java]: ditto

File: src/main/java/com/baidu/hugegraph/manager/RestoreManager.java [hunk 14]
```diff
@@ -222,11 +232,9 @@ private void restoreIndexLabels(HugeType type, String dir) {
...
     private void restore(HugeType type, File file,
                          BiConsumer<String, String> consumer) {
-        E.checkArgument(
-                file.exists() && file.isFile() && file.canRead(),
-                "Need to specify a readable filter file rather than: %s",
-                file.toString());
-
+        E.checkArgument(file.exists() && file.isFile() && file.canRead(),
+                        "Need to specify a readable filter file rather than:" +
+                        " %s", file.toString());
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/RestoreManager.java]: unneeded to call toString()

File: src/main/java/com/baidu/hugegraph/manager/RestoreManager.java [hunk 15]
```diff
@@ -240,35 +248,43 @@ private void restore(HugeType type, File file,
...
-            } else {
-                JavaType t = mapper.getTypeFactory()
-                                   .constructParametricType(List.class, clazz);
-                return (List<T>) mapper.readValue(element.toString(), t);
+    private void restoreZip(HugeType type, File file,
+                            BiConsumer<String, String> consumer) {
+        E.checkArgument(
+                file.exists() && file.isFile() && file.canRead(),
+                "Need to specify a readable filter file rather than: %s",
+                file.toString());
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/RestoreManager.java]: ditto

File: src/main/java/com/baidu/hugegraph/manager/TasksManager.java [hunk 16]
```diff
@@ -43,13 +43,13 @@
 
     private static long TASK_LIMIT = 10000;
 
-    public TasksManager(String url, String graph) {
-        super(url, graph, "tasks");
+    public TasksManager(String url, String graph, Integer timeout) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/TasksManager.java]: ditto

File: src/main/java/com/baidu/hugegraph/manager/RestoreManager.java [hunk 17]
```diff
@@ -58,13 +56,13 @@
 
     private Map<String, Long> primaryKeyVLs = null;
 
-    public RestoreManager(String url, String graph) {
-        super(url, graph, "restore");
+    public RestoreManager(String url, String graph, Integer timeout) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/RestoreManager.java]: ditto

File: src/main/java/com/baidu/hugegraph/manager/RestoreManager.java [hunk 18]
```diff
@@ -58,13 +56,13 @@
...
+        super(url, graph, "restore", timeout);
     }
 
     public RestoreManager(String url, String graph,
-                          String username, String password) {
-        super(url, graph, username, password, "restore");
+                          String username, String password, Integer timeout) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/RestoreManager.java]: ditto

File: src/main/java/com/baidu/hugegraph/manager/TasksManager.java [hunk 19]
```diff
@@ -43,13 +43,13 @@
...
+        super(url, graph, "tasks", timeout);
     }
 
     public TasksManager(String url, String graph,
-                        String username, String password) {
-        super(url, graph, username, password, "tasks");
+                        String username, String password, Integer timeout) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/TasksManager.java]: ditto

File: src/main/java/com/baidu/hugegraph/base/RetryManager.java [hunk 20]
```diff
@@ -30,39 +31,41 @@
...
+    public RetryManager(String url, String graph,
+                        String type, int timeout) {
+        super(url, graph, type, timeout);
     }
 
-    public RetryManager(String url, String graph,
-                        String username, String password, String type) {
-        super(url, graph, username, password, type);
+    public RetryManager(String url, String graph, String username,
+                        String password, String type, int timeout) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/base/RetryManager.java]: prefer add a class ConnectionInfo

**Review by javeme — CHANGES_REQUESTED** [review]

File: src/main/java/com/baidu/hugegraph/base/RetryManager.java [hunk 21]
```diff
@@ -85,18 +88,29 @@ public RetryManager(String url, String graph,
...
+                } catch (InterruptedException | ExecutionException e) {
+                    e.printStackTrace();
+                }
             }
-        }
+        } while (size < this.futures.size());
+        this.clearFutures();
+    }
+
+    private synchronized void clearFutures() {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/base/RetryManager.java]: remove clearFutures() and change futures to Collections.synchronizedList
also remove synchronized at line 91

File: src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java [hunk 22]
```diff
@@ -0,0 +1,251 @@
...
+            if (file.startsWith(type.string())) {
+                files.add(file);
+            }
+        }
+        return files;
+    }
+
+    protected String suffix() {
+        return this.operator.suffix();
+    }
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java]: remove this method just with one line

File: src/main/java/com/baidu/hugegraph/base/RetryManager.java [hunk 23]
```diff
@@ -20,49 +20,29 @@
...
             Executors.newFixedThreadPool(threadsNum);
-    private List<Future<?>> futures = new ArrayList<>();
-
-    protected static final long SPLIT_SIZE = 1024 * 1024L;
-    protected static final int LBUF_SIZE = 1024;
-
-    public RetryManager(String url, String graph, String type) {
-        super(url, graph, type);
-    }
+    private List<Future<?>> futures =
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/base/RetryManager.java]: set "private final", same as "private ExecutorService pool"

File: src/main/java/com/baidu/hugegraph/base/DirectoryOperator.java [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+        this.directory = directory;
+    }
+
+    public String directory() {
+        return this.directory;
+    }
+
+    public abstract List<String> files();
+
+    public String suffix() {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/base/DirectoryOperator.java]: set abstract

File: src/main/java/com/baidu/hugegraph/base/DirectoryOperator.java [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+
+package com.baidu.hugegraph.base;
+
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+
+import com.baidu.hugegraph.util.E;
+
+public abstract class DirectoryOperator {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/base/DirectoryOperator.java]: rename to Directory, same as subclass

File: src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java [hunk 26]
```diff
@@ -0,0 +1,251 @@
...
+    protected AtomicLong edgeLabelCounter = new AtomicLong(0);
+    protected AtomicLong indexLabelCounter = new AtomicLong(0);
+    protected AtomicLong vertexCounter = new AtomicLong(0);
+    protected AtomicLong edgeCounter = new AtomicLong(0);
+
+    private long startTime = 0L;
+    protected static KeyLock locks = new KeyLock();
+    private String logDir;
+    private Map<String, String> hdfsConf = new HashMap<>();
+    private DirectoryOperator operator;
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java]: rename operator to directory

**Review by javeme — CHANGES_REQUESTED** [review]

File: src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java [hunk 27]
```diff
@@ -0,0 +1,247 @@
...
+    public void init(SubCommands.BackupRestore cmd) {
+        assert cmd.retry() > 0;
+        this.retry(cmd.retry());
+        ensureDirectoryExist(cmd.logDir());
+        this.logDir(cmd.logDir());
+        String directory = cmd.directory();
+        E.checkArgument(directory != null && !directory.isEmpty(),
+                        "Directory of backup/restore can't be null or empty");
+        this.hdfsConf = cmd.hdfsConf();
+        if (hdfsDir(directory)) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java]: directory.startsWith(HdfsDirectory.HDFS_PREFIX)

File: src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java [hunk 28]
```diff
@@ -0,0 +1,247 @@
...
+                .put("edge label number", this.edgeLabelCounter.longValue())
+                .put("index label number", this.indexLabelCounter.longValue())
+                .put("vertex number", this.vertexCounter.longValue())
+                .put("edge number", this.edgeCounter.longValue()).build();
+        Printer.printMap(type + " summary", summary);
+
+        Printer.printKV("cost time(s)", this.elapseSeconds());
+    }
+
+    protected static void ensureDirectoryExist(String dir) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java]: unused method?

**Review by javeme — CHANGES_REQUESTED** [review]

File: src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java (hunk 28 above)
Comment (zhoney) [reply src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java]: used by ensure log dir in local fs exist

File: src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java [hunk 29]
```diff
@@ -0,0 +1,243 @@
...
+    public void init(SubCommands.BackupRestore cmd) {
+        assert cmd.retry() > 0;
+        this.retry(cmd.retry());
+        ensureDirectoryExist(cmd.logDir());
+        this.logDir(cmd.logDir());
+        String directory = cmd.directory();
+        E.checkArgument(directory != null && !directory.isEmpty(),
+                        "Directory of backup/restore can't be null or empty");
+        this.hdfsConf = cmd.hdfsConf();
+        if (directory.startsWith(HDFS_PREFIX)) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java:88]: move to HdfsDirectory.HDFS_PREFIX

File: src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java (hunk 28 above)
Comment (javeme) [reply src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java]: ok, it can be moved to LocalDirectory or Util

**Review by javeme — CHANGES_REQUESTED** [review]

File: src/main/java/com/baidu/hugegraph/base/HdfsDirectory.java [hunk 30]
```diff
@@ -0,0 +1,165 @@
...
+        try {
+            if (override) {
+                os = fs.create(dest, true);
+            } else {
+                os = fs.append(dest);
+            }
+            zos = new ZipOutputStream(os);
+            ZipEntry entry = new ZipEntry(file);
+            zos.putNextEntry(entry);
+        } catch (IOException e) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/base/HdfsDirectory.java:151]: close os

File: src/main/java/com/baidu/hugegraph/base/LocalDirectory.java [hunk 31]
```diff
@@ -0,0 +1,127 @@
...
+    @Override
+    public InputStream inputStream(String file) {
+        String path = Paths.get(this.directory(), file).toString();
+        ZipInputStream zis;
+        try {
+            InputStream is = new FileInputStream(path);
+            zis = new ZipInputStream(is);
+            E.checkState(zis.getNextEntry() != null,
+                         "Invalid zip file '%s'", file);
+        } catch (IOException e) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/base/LocalDirectory.java]: close InputStream is

File: src/main/java/com/baidu/hugegraph/base/LocalDirectory.java [hunk 32]
```diff
@@ -0,0 +1,127 @@
...
+    public OutputStream outputStream(String file, boolean override) {
+        String path = Paths.get(this.directory(), file).toString();
+        FileOutputStream fos;
+        ZipOutputStream zos;
+        try {
+            fos = new FileOutputStream(path, !override);
+            zos = new ZipOutputStream(fos);
+            ZipEntry entry = new ZipEntry(file);
+            zos.putNextEntry(entry);
+        } catch (IOException e) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/base/LocalDirectory.java:101]: ditto

File: src/main/java/com/baidu/hugegraph/cmd/HugeGraphCommand.java [hunk 33]
```diff
@@ -241,33 +255,31 @@ private void checkMainParams() {
...
-                    return clz.getConstructor(String.class, String.class,
-                                              String.class, String.class)
-                              .newInstance(this.url(), this.graph(),
-                                           this.username(), this.password());
-                } else {
-                    return clz.getConstructor(String.class, String.class)
-                              .newInstance(this.url(), this.graph());
-                }
-            }
+            ToolClient.ConnectionInfo info = new ToolClient.ConnectionInfo(
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/cmd/HugeGraphCommand.java]: import ConnectionInfo

File: src/main/java/com/baidu/hugegraph/base/RetryManager.java [hunk 34]
```diff
@@ -90,13 +70,20 @@ public void submit(Runnable task) {
...
-        for (Future<?> future : this.futures) {
-            try {
-                future.get();
-            } catch (InterruptedException | ExecutionException e) {
-                e.printStackTrace();
+        int size = 0;
+        int offset;
+        do {
+            offset = size;
+            size = this.futures.size();
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/base/RetryManager.java]: change futures to use queue instead list

**Review by javeme — CHANGES_REQUESTED** [review]

File: src/main/java/com/baidu/hugegraph/base/LocalDirectory.java [hunk 35]
```diff
@@ -73,13 +73,21 @@ public String suffix() {
...
-            InputStream is = new FileInputStream(path);
+            is = new FileInputStream(path);
             zis = new ZipInputStream(is);
             E.checkState(zis.getNextEntry() != null,
                          "Invalid zip file '%s'", file);
         } catch (IOException e) {
+            if (is != null) {
+                try {
+                    is.close();
+                } catch (Exception e1) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/base/LocalDirectory.java]: Exception ignored

File: src/main/java/com/baidu/hugegraph/base/HdfsDirectory.java [hunk 36]
```diff
@@ -128,6 +128,13 @@ public InputStream inputStream(String file) {
             E.checkState(zis.getNextEntry() != null,
                          "Invalid zip file '%s'", file);
         } catch (IOException e) {
+            if (is != null) {
+                try {
+                    is.close();
+                } catch (Exception e1) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/base/HdfsDirectory.java]: use `Exception ignored`
and add a method closeAndIgnoreError()

**Review by javeme — CHANGES_REQUESTED** [review]

File: src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java [hunk 37]
```diff
@@ -0,0 +1,232 @@
...
+            throw new ToolsException("Failed to deserialize %s from %s",
+                                     e, type, file);
+        }
+    }
+
+    private OutputStream outputStream(String file) {
+        OutputStream os;
+        if (this.outputStreams.get(file) == null) {
+            os = this.directory.outputStream(file, true);
+            this.outputStreams.putIfAbsent(file, os);
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java]: close os if putIfAbsent does not return null

File: src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java [hunk 38]
```diff
@@ -0,0 +1,232 @@
...
+        }
+        return this.outputStreams.get(file);
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java]:
```
        OutputStream os = this.outputStreams.get(file);
        if (os != null) {
            return os;
        }
        os = this.directory.outputStream(file, true);
        OutputStream prev = this.outputStreams.putIfAbsent(file, os);
        if (prev != null) {
            closeAndIgnore(os);
            os = prev;
        }
        return os;
```

File: src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java [hunk 39]
```diff
@@ -0,0 +1,232 @@
...
+    }
+
+    private InputStream inputStream(String file) {
+        InputStream is;
+        if (this.inputStreams.get(file) == null) {
+            is = this.directory.inputStream(file);
+            this.inputStreams.putIfAbsent(file, is);
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java]: ditto

File: src/main/java/com/baidu/hugegraph/base/LocalDirectory.java [hunk 40]
```diff
@@ -0,0 +1,131 @@
...
+
+    @Override
+    public InputStream inputStream(String file) {
+        String path = Paths.get(this.directory(), file).toString();
+        InputStream is = null;
+        ZipInputStream zis;
+        try {
+            is = new FileInputStream(path);
+            zis = new ZipInputStream(is);
+            E.checkState(zis.getNextEntry() != null,
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/base/LocalDirectory.java:81]: this will cause InputStream leak

**Review by javeme — CHANGES_REQUESTED** [review]

File: src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java [hunk 41]
```diff
@@ -0,0 +1,242 @@
...
+                .put("edge label number", this.edgeLabelCounter.longValue())
+                .put("index label number", this.indexLabelCounter.longValue())
+                .put("vertex number", this.vertexCounter.longValue())
+                .put("edge number", this.edgeCounter.longValue()).build();
+        Printer.printMap(type + " summary", summary);
+
+        Printer.printKV("cost time(s)", this.elapseSeconds());
+    }
+
+    protected void write(String file, String type, List<?> list) {
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/BackupRestoreBaseManager.java]: what is the difference between this function and the previous one at line 109?

File: src/main/java/com/baidu/hugegraph/manager/RestoreManager.java [hunk 42]
```diff
@@ -217,64 +213,13 @@ private void restoreIndexLabels(HugeType type, String dir) {
                 this.indexLabelCounter.getAndIncrement();
             }
         };
-        this.restore(type, Paths.get(dir, fileName).toFile(), consumer);
+        String path = this.filesWithPrefix(HugeType.INDEX_LABEL).get(0);
+        this.restore(type, path, consumer);
     }
 
-    private void restore(HugeType type, File file,
+    private void restore(HugeType type, String file,
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/RestoreManager.java:220]: seems restore() can be deleted

File: src/main/java/com/baidu/hugegraph/base/HdfsDirectory.java [hunk 43]
```diff
@@ -0,0 +1,168 @@
...
+
+import com.baidu.hugegraph.exception.ToolsException;
+import com.baidu.hugegraph.rest.ClientException;
+import com.baidu.hugegraph.util.E;
+
+public class HdfsDirectory extends Directory {
+
+    public static final String HDFS_PREFIX = "hdfs://";
+
+    private Map<String, String> conf;
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/base/HdfsDirectory.java]: set final

File: src/main/java/com/baidu/hugegraph/base/Directory.java [hunk 44]
```diff
@@ -0,0 +1,67 @@
...
+import java.util.List;
+
+import com.baidu.hugegraph.util.E;
+
+public abstract class Directory {
+
+    private String directory;
+
+    public Directory(String directory) {
+        E.checkArgument(directory != null && !directory.isEmpty(),
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/base/Directory.java:34]: move the check to directory(String directory) and call directory(directory)

File: src/main/java/com/baidu/hugegraph/manager/RestoreManager.java [hunk 45]
```diff
@@ -172,11 +168,11 @@ private void restorePropertyKeys(HugeType type, String dir) {
                 this.propertyKeyCounter.getAndIncrement();
             }
         };
-        this.restore(type, Paths.get(dir, fileName).toFile(), consumer);
+        String path = this.filesWithPrefix(HugeType.PROPERTY_KEY).get(0);
```
Comment (javeme) [inline src/main/java/com/baidu/hugegraph/manager/RestoreManager.java]: add a method fileWithPrefix() to return one path, and check just one

**Review by javeme — CHANGES_REQUESTED** [review]

**Review by javeme — COMMENTED** [review]
LGTM

File: src/main/java/com/baidu/hugegraph/cmd/SubCommands.java [hunk 46]
```diff
@@ -390,19 +384,43 @@ public boolean force() {
 
     public class BackupRestore {
 
+        @Parameter(names = {"--directory", "-d"}, arity = 1,
+                   description = "Directory of graph schema/data")
+        public String directory = "./";
+
+        @Parameter(names = {"--log", "-l"}, arity = 1,
+                description = "Directory of log")
```
Comment (Linary) [inline src/main/java/com/baidu/hugegraph/cmd/SubCommands.java]: align

File: src/main/java/com/baidu/hugegraph/base/HdfsDirectory.java [hunk 47]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    private FileSystem fileSystem() {
+        Configuration conf = new Configuration();
+        for (Map.Entry<String, String> entry : this.conf.entrySet()) {
+            conf.set(entry.getKey(), entry.getValue());
+        }
+        FileSystem fs;
+        try {
+            fs = FileSystem.get(conf);
```
Comment (Linary) [inline src/main/java/com/baidu/hugegraph/base/HdfsDirectory.java]: just return FileSystem.get(conf);

File: src/main/java/com/baidu/hugegraph/base/HdfsDirectory.java [hunk 48]
```diff
@@ -0,0 +1,168 @@
...
+        } catch (IOException e) {
+            throw new ClientException("Failed to access HDFS with " +
+                                      "configuration %s", this.conf, e);
+        }
+        return fs;
+    }
+
+    @Override
+    public List<String> files() {
+        List<String> files = new ArrayList<>();
```
Comment (Linary) [inline src/main/java/com/baidu/hugegraph/base/HdfsDirectory.java]: suggest move to line 79

**Review by Linary — CHANGES_REQUESTED** [review]

**Review by Linary — APPROVED** [review]

**Review by javeme — APPROVED** [review]

**Pull request merged**